Add initial linux_dmabuf protocol support

agx opened the pull request:
Tested with ./weston-simple-dmabuf-drm
  
Todo  
- [x]  fix FIXMEs
- [x] style cleanups
- [x] check build without drm headers
- [x] more tests
- [x] drop renderer indirection
- [x] check error handling (logged errors vs. errors reported to clients)
 -later:  look at YUV
Initial feedback would be welcome.

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+
+static void wl_buffer_destroy(struct wl_client *client,
+	struct wl_resource *resource)
+{
```
Style nitpick: no newlines between `)` and `{`

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+		return NULL;
+	}
+
+	buffer = wl_resource_get_user_data(resource);
```
* We generally create `<type>_from_resource` helpers to provide better type safety (for instance, https://github.com/swaywm/wlroots/blob/master/types/wlr_gamma_control.c#L28). You'll be able to re-use this helper on line 64.
* Style nitpick: declare variables when they're first used

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	buffer = wl_resource_get_user_data(resource);
+	assert(buffer);
```
This is not necessary.

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+}
+
+static void
+params_destroy(struct wl_client *client, struct wl_resource *resource) {
```
Style nitpick: no newline between `static void` and the rest of the line

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	assert(buffer->params_resource == params_resource);
+	assert(!buffer->buffer_resource);
```
These assertions could be moved in the `*_from_resource` helper

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+
+	/* initial buffer import */
+	if (!wlr_renderer_import_dmabuf(buffer->renderer, buffer))
+		goto err_failed;
```
Style nitpick: always use brackets

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,431 @@
+	return;
+
+err_failed:
+	if (buffer_id == 0)
```
Style nitpick: use brackets

emersion commented on render/wlr_renderer.c:
```diff
@@ -72,3 +72,17 @@ bool wlr_renderer_format_supported(struct wlr_renderer *r,
+
+bool wlr_renderer_import_dmabuf(struct wlr_renderer *r,
+	  struct wlr_dmabuf_buffer *dmabuf) {
+	return r->impl->import_dmabuf(r, dmabuf);
```
Check if this function is NULL, return false if so?

emersion commented on include/wlr/types/wlr_linux_dmabuf.h:
```diff
@@ -0,0 +1,42 @@
+	uint32_t offset[MAX_LINUX_BUFFER_PLANES];
+	uint32_t stride[MAX_LINUX_BUFFER_PLANES];
+	uint64_t modifier[MAX_LINUX_BUFFER_PLANES];
+	int fd[MAX_LINUX_BUFFER_PLANES];
```
You could maybe instead use a struct for `offset`, `stride`, `modifier` and `fd`?

ascent12 replied:
It's much more consistent for other dmabuf APIs if it stays this way.

SirCmpwn commented:
Looking good so far. Take care to review the style guide, please.

ascent12 commented:
I would prefer if this was in terms of wlr_egl instead of wlr_renderer.
All of the gles2 dmabuf functions are just calling EGL functions. It's already capable of importing EGLImages, so we can just leave it at that. wlr_renderer is "supposed" (I don't think it actually is right now) to be an optional type, so I don't want it tangled with any of our other interfaces.

My work on Vulkan will certainly rethinking part of this interface, and its relationship between the renderer, but that work probably won't land any time soon, so it shouldn't be considered a blocker.

agx commented:
On Fri, Mar 02, 2018 at 11:26:03AM -0800, Scott Anderson wrote:
> I would prefer if this was in terms of wlr_egl instead of
> wlr_renderer.  All of the gles2 dmabuf functions are just calling EGL
> functions. It's already capable of importing EGLImages, so we can just
> leave it at that. wlr_renderer is "supposed" (I don't think it
> actually is right now) to be an optional type, so I don't want it
> tangled with any of our other interfaces.

I only took the detour via wlr_renderer since no other code in types/
used wlr_egl_* directly but wlr_render{,er}_*. If it's o.k. to do so I'm
happy to drop the indirection.

> My work on Vulkan will certainly rethinking part of this interface,
> and its relationship between the renderer, but that work probably
> won't land any time soon, so it shouldn't be considered a blocker.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/swaywm/wlroots/pull/698#issuecomment-370026418

ascent12 commented:
Yeah, the relationship between wlr_egl and wlr_renderer isn't very obvious and not spelled out anywhere.
wlr_egl is geared towards integrating with the platform/backends, while wlr_renderer is just drawing stuff.
Importing dmabufs fits into 'integrating with the platform'.

The idea was that a library user could use wlr_egl with their own renderer code if they wanted to do something more advanced.

agx commented:
@ascent12 Thanks for spelling this out so clearly. I'll drop the renderer indirection.

emersion commented on rootston/desktop.c:
```diff
@@ -729,9 +730,9 @@ struct roots_desktop *desktop_create(struct roots_server *server,
-	desktop->linux_dmabuf = wlr_linux_dmabuf_create(
-		server->wl_display, server->renderer);
 
+	struct wlr_gles2_renderer *r = (struct wlr_gles2_renderer *)server->renderer;
```
You can use `wlr_backend_get_egl` instead of casting the renderer.

emersion commented on include/wlr/render/matrix.h:
```diff
@@ -8,6 +8,7 @@ void wlr_matrix_identity(float (*output)[16]);
 void wlr_matrix_translate(float (*output)[16], float x, float y, float z);
 void wlr_matrix_scale(float (*output)[16], float x, float y, float z);
 void wlr_matrix_rotate(float (*output)[16], float radians);
+void wlr_matrix_invert_y(float (*output)[16]);
```
Seems that this one isn't relevant anymore.

agx commented:
On Wed, Mar 07, 2018 at 08:50:55AM -0800, emersion wrote:
> emersion commented on this pull request.
> 
> 
> 
> > @@ -8,6 +8,7 @@ void wlr_matrix_identity(float (*output)[16]);
>  void wlr_matrix_translate(float (*output)[16], float x, float y, float z);
>  void wlr_matrix_scale(float (*output)[16], float x, float y, float z);
>  void wlr_matrix_rotate(float (*output)[16], float radians);
> +void wlr_matrix_invert_y(float (*output)[16]);
> 
> Seems that this one isn't relevant anymore.

The commit message wasn't nice either. Thanks for having a look!

SirCmpwn commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+
+	uint32_t version = wl_resource_get_version(linux_dmabuf_resource);
+	struct wlr_dmabuf_buffer *buffer = calloc(1, sizeof *buffer);
+	if (!buffer)
```
Braces mandatory

SirCmpwn commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+		goto err_out;
+	}
+
+	if ((uint64_t) buffer->attributes.offset[0] +
```
No space after cast

SirCmpwn commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+}
+
+struct wlr_linux_dmabuf *wlr_linux_dmabuf_create(struct wl_display *display,
+												 struct wlr_egl *egl) {
```
Not sure what happened here

SirCmpwn commented:
Style issues aside this LGTM

agx commented:
@SirCmpwn thanks. I've cleaned up these style issues plus some more and shuffeled the patches around grouped by functionality.

emersion commented on include/wlr/render.h:
```diff
@@ -72,6 +73,14 @@ bool wlr_renderer_read_pixels(struct wlr_renderer *r, enum wl_shm_format fmt,
+
+int wlr_renderer_get_dmabuf_formats(struct wlr_renderer *r, int **formats);
+int wlr_renderer_get_dmabuf_modifiers(struct wlr_renderer *r, int format,
+	uint64_t **modifiers);
```
These three functions don't exist anymore.

emersion commented on include/wlr/types/wlr_linux_dmabuf.h:
```diff
@@ -0,0 +1,73 @@
+#ifndef WLR_TYPES_WLR_SIMPLE_DMABUF_H
+#define WLR_TYPES_WLR_SIMPLE_DMABUF_H
+
+#define MAX_LINUX_BUFFER_PLANES 4
```
This should be prefixed by `WLR_LINUX_DMABUF` since this is exported.

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+
+static void
+handle_buffer_destroy(struct wl_resource *buffer_resource)
+{
```
Style nit: should be all three on the same line

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+};
+
+struct wlr_dmabuf_buffer *wlr_dmabuf_buffer_from_params_resource(
+	struct wl_resource *params_resource) {
```
Style nit: missing an extra tab here, so we see the difference between the function signature and the function body.

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+static void handle_params_destroy(struct wl_resource *params_resource) {
+	struct wlr_dmabuf_buffer *buffer = wl_resource_get_user_data(params_resource);
+	/* Check for NULL since wlr_dmabuf_buffer_from_params_resource will choke */
+	if (!buffer) {
```
I'd prefer not casting here but rather write:

```c
if (!wl_resource_get_user_data(params_resource)) {
```

So we can't by mistake use the unchecked `struct wlr_dmabuf_buffer *`

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+static void linux_dmabuf_create_params(struct wl_client *client,
+		struct wl_resource *linux_dmabuf_resource,
+		uint32_t params_id) {
+	struct wlr_linux_dmabuf *linux_dmabuf = wl_resource_get_user_data(linux_dmabuf_resource);
```
There should be a `wlr_linux_dmabuf_from_resource`

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+		goto err;
+	}
+
+	for (int i = 0; i < MAX_LINUX_BUFFER_PLANES; i++)
```
Style nit: brackets wanted!

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+	buffer->params_resource = wl_resource_create(client,
+		&zwp_linux_buffer_params_v1_interface,
+		version, params_id);
+	if (!buffer->params_resource)
```
Style nit: ditto

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+
+static const struct zwp_linux_dmabuf_v1_interface linux_dmabuf_impl = {
+        linux_dmabuf_destroy,
+        linux_dmabuf_create_params
```
Style nit: this isn't a tab

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+				modifier_hi,
+				modifier_lo);
+		}
+		if (modifiers != &modifier_invalid)
```
Style nit: brackets

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,435 @@
+
+	wl_resource_set_implementation(resource, &linux_dmabuf_impl,
+		linux_dmabuf, NULL);
+	if (version < ZWP_LINUX_DMABUF_V1_MODIFIER_SINCE_VERSION)
```
Style nit: brackets

agx commented:
Besides the style fixes (@emersion thanks for review) I've moved the check for YUV surfaces from gles2_texture_upload_dmabuf to wlr_egl_check_import_dmabuf so the client knows that we can't render the texture instead of just having a warning in the logs

emersion commented on include/wlr/render/egl.h:
```diff
@@ -61,6 +64,31 @@ EGLSurface wlr_egl_create_surface(struct wlr_egl *egl, void *window);
 
+/**
+ * Creates an egl image from the given dmabuf attributes. Check usability
+ * of the dmabuf with egl_check_import_dmabuf once first.
```
Nitpick: should be `with wlr_egl_check_import_dmabuf`

emersion commented on include/wlr/render/egl.h:
```diff
@@ -61,6 +64,31 @@ EGLSurface wlr_egl_create_surface(struct wlr_egl *egl, void *window);
+/**
+ * Get the available dmabuf formats
+ */
+int wlr_egl_get_dmabuf_formats(struct wlr_egl *egl,	int **formats);
```
Nitpick: here's an extra tab here

emersion commented on include/wlr/types/wlr_linux_dmabuf.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef WLR_TYPES_WLR_SIMPLE_DMABUF_H
+#define WLR_TYPES_WLR_SIMPLE_DMABUF_H
```
Nitpick: should be `WLR_TYPES_WLR_LINUX_DMABUF_H`

emersion commented on include/wlr/types/wlr_linux_dmabuf.h:
```diff
@@ -0,0 +1,80 @@
+ * Create linux-dmabuf interface
+ */
+struct wlr_linux_dmabuf *wlr_linux_dmabuf_create(struct wl_display *display,
+												 struct wlr_egl *egl);
```
Nitpick: no need to indent the second line with so many tabs/spaces, one is enough.

emersion commented on types/wlr_linux_dmabuf.c:
```diff
@@ -0,0 +1,449 @@
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_ALREADY_USED,
+			"params was already used to create a wl_buffer");
+		return;
```
This branch is never reached, because `wlr_dmabuf_buffer_from_params_resource` asserts that the buffer isn't NULL.

A fix would be to replace the condition by `if (!wl_resource_get_user_data(params_resource)) {` and move the `wlr_dmabuf_buffer_from_params_resource` call below.

agx commented:
@emersion thanks for the review. I've fixed that and also plugged a resoure leak in wlr_egl_check_import_dmabuf where we I did not free the eglimage in case of a YUV format, dropped the unused user_data and fixed typo comments.

agx commented:
I have an additional patch that adds support for multiple planes:

     https://github.com/agx/wlroots/commit/a279202e648588bbe5fb23cc6b88a2b66d3beb24

shall I put that on on top or use a different PR? I can't really test this since I dont have NV12 support but it gets as far as with weston so it shouldn't make things worse.

emersion commented on types/wlr_surface.c:
```diff
@@ -325,6 +325,10 @@ static void wlr_surface_apply_damage(struct wlr_surface *surface,
 			wlr_texture_upload_drm(surface->texture, surface->current->buffer);
 			goto release;
+		} else if (wlr_dmabuf_buffer_from_buffer_resource(
+					   surface->current->buffer)) {
```
`wlr_dmabuf_buffer_from_buffer_resource` aborts if the resource isn't a dmabuf buffer or if the buffer is NULL.

You seem to assume here (and in other places too, grep for `wlr_dmabuf_buffer_from_buffer_resource`) that it returns NULL instead of aborting. Maybe a `wlr_dmabuf_resource_is_buffer` could be used instead?

agx commented:
@emersion a leftover from when the ``wlr_dmabuf_buffer_from_buffer_resource`` didn't have the asserts. ``wlr_dmabuf_resource_is_buffer`` sounds good so I added it.

emersion commented:
Thanks!

>I have an additional patch that adds support for multiple planes

Let's do this in a new PR, make sure to provide a test plan!

SirCmpwn commented:
Thanks!

agx commented:
That was https://github.com/swaywm/wlroots/pull/698/commits/2d0db169422b74e5ef7c07965c376273c7bb2961 in particular. Thanks for fixing this!
